support authentication via oauth2 personal access tokens

ryanpetrello opened the pull request:
see: https://github.com/ansible/tower-cli/issues/478

```
$ tower-cli config host tower.example.org
Configuration updated successfully.

$ tower-cli login bob
Password:
Configuration updated successfully.

$ cat ~/.tower_cli.cfg
[general]
host = tower.example.org
oauth_token = 6LvR6QhttzU2cpIBHc2uOqMbCEbPcH

$ tower-cli inventory list
== ============== ============
id      name      organization
== ============== ============
 1 Demo Inventory            1
== ============== ============
```

ryanpetrello commented:
In another PR, I'd like to extend this new `login` command to support app-based token generation, e.g.,

`$ tower-cli login bob --client abc --secret xyz`

rooftopcellist commented on tests/test_cli_misc.py:
```diff
@@ -269,6 +272,106 @@ def test_error_no_global_config_file(self):
+                    t._registry[six.text_type('{}o/').format(prefix)] = Resp(
+                        '', 200, {}
+                    )
+                    t.register('/users/bob/personal_tokens/', json.dumps({
```
I'm curious, is it ok to have 'bob' there instead of the user's pk?

ryanpetrello replied:
Named URLs :D

AlanCoding commented on tower_cli/cli/misc.py:
```diff
@@ -204,3 +205,30 @@ def config(key=None, value=None, scope='user', global_=False, unset=False):
+
+    if r.status_code == 201:
+        token = r.json()['token']
+        config.main(['oauth_token', token, '--scope=user'])
```
Well this is a very interesting approach.

ryanpetrello replied:
```shell
~/dev/tower-cli tower-cli config

# User options (set with `tower-cli config`; stored in ~/.tower_cli.cfg).
host: tower.example.org:8043

# Defaults.
username:
use_token: False
verbose: False
certificate:
format: human
color: True
description_on: False
verify_ssl: True
oauth_token:
password:

~/dev/tower-cli tower-cli inventory list --tower-oauth-token=PB3X8L2hXjN4mxdYxt7dd0PbInhSDk
== ============== ============
id      name      organization
== ============== ============
 1 Demo Inventory            1
== ============== ============
```

AlanCoding commented:
So is `oauth_token` shown as a global option for all commands? Does it work when used that way? (getting it from the runtime settings)

Can you set it with an env var?

ryanpetrello commented:
I'm super excited about the fact that these tests pass locally for me :)

ryanpetrello commented:
@AlanCoding 

> So is oauth_token shown as a global option for all commands? Does it work when used that way? (getting it from the runtime settings)

With the latest code in this PR:

```shell
~/dev/tower-cli tower-cli config

# User options (set with `tower-cli config`; stored in ~/.tower_cli.cfg).
host: tower.example.org:8043

# Defaults.
username:
use_token: False
verbose: False
certificate:
format: human
color: True
description_on: False
verify_ssl: True
oauth_token:
password:

~/dev/tower-cli tower-cli inventory list --tower-oauth-token=PB3X8L2hXjN4mxdYxt7dd0PbInhSDk
== ============== ============
id      name      organization
== ============== ============
 1 Demo Inventory            1
== ============== ============
```

AlanCoding commented:
First-world problems: the variables are showing up in defaults even if I haven't set them.

```
$ tower-cli config

# User options (set with `tower-cli config`; stored in ~/.tower_cli.cfg).
username: admin
host: http://localhost:8013/
verify_ssl: False
verbose: False

# Defaults.
use_token: False
certificate: 
format: human
color: True
description_on: False
oauth_token: 
password: 
```

There's got to be some way to recognize a null-ish state, because presumably that would just be copying the implementation for all the other settings.

AlanCoding commented:
If I can `tower-cli login admin`, then I feel like I should be able to `tower-cli logout admin`

AlanCoding commented:
It takes an uncomfortably long amount of time after I type my password and press enter for the command to finish. It makes you think that you did something wrong. I think it would be good to give a message like "waiting for carrier pidgins to return"

AlanCoding commented:
I also don't have a verbose option to debug what requests are being made. We don't have that option for config, but no requests are being made for that.

AlanCoding commented:
should logging out actually reach out and invalidate the token?

AlanCoding commented:
the response gives

```json
"expires": "2018-03-07T23:02:15.648450Z", 
```

That would be great information to pass along at the end of the command!

rooftopcellist commented on tower_cli/cli/misc.py:
```diff
@@ -204,3 +205,30 @@ def config(key=None, value=None, scope='user', global_=False, unset=False):
+        )
+
+    # Explicitly set a basic auth header for PAT acquisition (so that we don't
+    # try to auth w/ an existing user+pass or oauth token in a config file)
```
We may want to change `oauth` to `oauth2`for consistency with AWX.

rooftopcellist commented:
@AlanCoding -- Re:
> should logging out actually reach out and invalidate the token?

Can the user continue to use this token to authenticate after changing tower-cli tower A to tower B, then back to A?  

If I understand this right: on `tower-cli login admin`, it passes user/pass to acquire a token, then uses that token for all future tower-cli requests.  If this token is only used for tower-cli, it would be ok to revoke on `tower-cli logout admin`.  

My worry is that a consumer of tower may see the token and try to use it for something else, then be disappointed when the token disappeared from a seemingly unrelated tower-cli logout.  Solution: we could add a warning message on login and logout.  (or just logout)

AlanCoding commented:
Those are good points, and right now I tend to agree with this behavior

> If this token is only used for tower-cli, it would be ok to revoke on `tower-cli logout admin`.  

In the logout help text, we should offer `tower-cli config oauth_token --unset` as an alternative that keeps the token active.

ryanpetrello commented:
@AlanCoding @rooftopcellist this PR is getting pretty tall.  I've addressed:

- add a `-v` flag for `tower-cli login`
- print the response JSON (including token value and expiration) when `tower-cli login` succeeds
- added a `--scope` flag, which takes `'read'` and `'write'`, but which defaults to `'write'`

Would it be okay if I added `tower-cli logout` in a separate PR?

rooftopcellist commented:
I think splitting it up would be fine in this case.  @AlanCoding So to be clear, the --unset flag would just make tower-cli use basic auth, but still keep the token stored, right?

ryanpetrello commented on tests/test_api.py:
```diff
@@ -248,7 +249,9 @@ def test_reading_expired_token(self):
                     t.register('/authtoken/', json.dumps({'token': 'barfoo', 'expires': expires}), status_code=200,
                                method='POST')
-                    self.auth(self.req)
+                    with warnings.catch_warnings():
```
Tests were noisy without this.

ryanpetrello commented on tests/test_api.py:
```diff
@@ -269,5 +272,16 @@ def test_auth_token_unsupported(self):
                 t.register('/authtoken/', json.dumps({}), status_code=404, method='OPTIONS')
                 auth = BasicTowerAuth('alice', 'pass', client)
-                auth(self.req)
+                with warnings.catch_warnings():
```
Same - noisy tests.

ryanpetrello commented on tower_cli/api.py:
```diff
@@ -156,8 +166,7 @@ def _make_request(self, method, url, args, kwargs):
 
-    @property
-    def prefix(self):
+    def get_prefix(self, include_version=True):
```
Up until now, every endpoint we've wanted to hit has been prefixed with a version, so I had to change this from a property to a function for generating `/api/o/` prefixes.

ryanpetrello commented on tower_cli/api.py:
```diff
@@ -223,12 +236,12 @@ def request(self, method, url, *args, **kwargs):
         # If so, fail out now; this is always a failure.
         if r.status_code == 401:
-            raise exc.AuthError('Invalid Tower authentication credentials.')
+            raise exc.AuthError('Invalid Tower authentication credentials (HTTP 401).')
```
This is beyond the scope of this issue, and PR, but I think it may take debugging a bit clearer for folks (at least it did for me when dealing w/ read vs. write oauth2 scopes).

ryanpetrello commented on tower_cli/cli/misc.py:
```diff
@@ -204,3 +207,39 @@ def config(key=None, value=None, scope='user', global_=False, unset=False):
+@click.option('--password', required=True, prompt=True, hide_input=True)
+@click.option('--scope', required=False, default='write',
+              type=click.Choice(['read', 'write']))
+@click.option('-v', '--verbose', default=None,
```
I opted to just specify `-v` manually instead of inheriting _all_ global options, because some of them don't really make sense for this command.

ryanpetrello commented on tower_cli/cli/misc.py:
```diff
@@ -204,3 +207,39 @@ def config(key=None, value=None, scope='user', global_=False, unset=False):
+
+    if r.status_code == 201:
+        result = r.json()
+        result.pop('summary_fields', None)
```
These two keys are a bit too verbose/unimportant for this tool.

ryanpetrello commented on tower_cli/cli/misc.py:
```diff
@@ -204,3 +207,39 @@ def config(key=None, value=None, scope='user', global_=False, unset=False):
+        result.pop('summary_fields', None)
+        result.pop('related', None)
+        token = result['token']
+        secho(json.dumps(result, indent=1), fg='blue', bold=True)
```
![screen shot 2018-03-07 at 4 45 52 pm](https://user-images.githubusercontent.com/214912/37119957-066c0bb8-2227-11e8-883c-e83dcf73cc1f.png)
